### ISSUE TITLE
ty reports `possibly-unresolved-reference` on a `for`-loop variable even if iteration is guaranteed

### ISSUE THREAD

**Issue by zeevro** [astral-sh/ty#2946]
### Summary

*`test.py`:*
```py
def f(l: list[int]) -> int:
    if not l:
        return 0
    reveal_type(l)
    for i in l:
        pass
    return i
```
```console
$ ty check test.py --error possibly-unresolved-reference --ignore undefined-reveal --output-format concise
test.py:4:17: info[revealed-type] Revealed type: `list[int] & ~AlwaysFalsy`
test.py:7:12: error[possibly-unresolved-reference] Name `i` used when possibly not defined
Found 2 diagnostics
```
[playground](https://play.ty.dev/5373741d-4ab8-42a1-9dfc-c1503bef7f0b)

### Version

ty 0.0.20

**Comment by AlexWaygood** [comment]
Thanks for the report. This is expected behaviour -- [as we document](https://docs.astral.sh/ty/reference/rules/#possibly-unresolved-reference), false positives like this are the reason why this rule is disabled by default; you have to explicitly opt into it if you want this very strict lint to fire on your project. It's true that if you know the instance of a class is exactly a `list` (and not a subclass of a `list`), and you know that the list is truthy, then you can say for sure that the loop will always run for >=1 iteration. But it's not true that for any arbitrary iterable, it will always run for >=1 iteration if it's truthy, and it's not even necessarily true for all lists that they will always run for >=1 iteration if they're truthy. (You could have a subclass of list that has custom behaviour in its `__bool__` method.)

For reference, both pyright and pycroscope emit the same diagnostic here (`reportPossiblyUnboundVariable` for pyright, `possibly_undefined_name` for pycroscope). Mypy and pyrefly don't seem to (that's true even if I run mypy with `--strict --enable-error-code=possibly-undefined`), but I'm not sure why not.
